pkg/controller/datavolume/upload-controller.go

```diff
@@ -163,3 +233,59 @@ func (r *UploadReconciler) updateStatusPhase(pvc *corev1.PersistentVolumeClaim,
 	}
 	return nil
 }
+
+func volumeUploadSourceName(dv *cdiv1.DataVolume) string {
+	return fmt.Sprintf("%s-%s", volumeUploadSourcePrefix, dv.UID)
+}
+
+func (r *UploadReconciler) createVolumeUploadSourceCR(syncState *dvSyncState) error {
+	uploadSourceName := volumeUploadSourceName(syncState.dvMutated)
+	uploadSource := &cdiv1.VolumeUploadSource{}
+	exists, err := cc.GetResource(context.TODO(), r.client, syncState.dvMutated.Namespace, uploadSourceName, uploadSource)
+	if err != nil {
+		return err
+	}
+	if exists {
+		// uploadSource already exists
+		return nil
+	}
+
+	uploadSource = &cdiv1.VolumeUploadSource{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      uploadSourceName,
+			Namespace: syncState.dv.Namespace,
+		},
+		Spec: cdiv1.VolumeUploadSourceSpec{
+			ContentType:   syncState.dv.Spec.ContentType,
+			Preallocation: syncState.dv.Spec.Preallocation,
+		},
+	}
+
+	if err := controllerutil.SetControllerReference(syncState.dvMutated, uploadSource, r.scheme); err != nil {
+		return err
+	}
+	if err := r.client.Create(context.TODO(), uploadSource); err != nil {
+		if !k8serrors.IsAlreadyExists(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *UploadReconciler) deleteVolumeUploadSourceCR(syncState *dvSyncState) error {
+	uploadSourceName := volumeUploadSourceName(syncState.dvMutated)
+	uploadSource := &cdiv1.VolumeUploadSource{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: uploadSourceName, Namespace: syncState.dvMutated.Namespace}, uploadSource); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+	} else {
+		if err := r.client.Delete(context.TODO(), uploadSource); err != nil {
+			if !k8serrors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
```

Comment:
Sama as before, `Get()` is not needed before `Delete()` and after removing it use `err != nil && !k8serrors.IsAlreadyExists(err)`

Reply:
Ill take the Get suggestion, I like better to have the not found err in another line :)